File: java/src/main/java/kata/DeliveryController.java
package kata;

import java.time.Duration;
import java.time.format.DateTimeFormatter;
import java.util.List;

public class DeliveryController {

  private final EmailGateway _emailGateway;
  private final MapService _mapService;
  public List<Delivery> deliverySchedule;

  public DeliveryController(List<Delivery> deliverySchedule) {
    this.deliverySchedule = deliverySchedule;
    this._emailGateway = new EmailGateway();
    this._mapService = new MapService();
  }

  public void updateDelivery(DeliveryEvent deliveryEvent) {
    Delivery nextDelivery = null;
    for (int i = 0; i < deliverySchedule.size(); i++) {
      Delivery delivery = deliverySchedule.get(i);
      if (deliveryEvent.id() == delivery.id()) {
        delivery.setArrived(true);
        Duration d = Duration.between(delivery.timeOfDelivery(), deliveryEvent.timeOfDelivery());

        if (d.toMinutes() < 10) {
          delivery.setOnTime(true);
        }
        delivery.setTimeOfDelivery(deliveryEvent.timeOfDelivery());
        String message = "Regarding your delivery today at %s. How likely would you be to recommend this delivery service to a friend? Click <a href='url'>here</a>".formatted(
            DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm").format(delivery.timeOfDelivery()));
        _emailGateway.send(delivery.contactEmail(), "Your feedback is important to us", message);
        if (deliverySchedule.size() > i + 1) {
          nextDelivery = deliverySchedule.get(i + 1);
        }

        if (!delivery.onTime() && deliverySchedule.size() > 1 && i > 0) {
          var previousDelivery = deliverySchedule.get(i - 1);
          Duration elapsedTime = Duration.between(previousDelivery.timeOfDelivery(),
              delivery.timeOfDelivery());
          _mapService.updateAverageSpeed(previousDelivery.location(), delivery.location(),
              elapsedTime);
        }
      }
    }

    if (nextDelivery != null) {
      var nextEta = _mapService.calculateETA(deliveryEvent.location(), nextDelivery.location());
      var message =

          "Your delivery to %s is next, estimated time of arrival is in %s minutes. Be ready!".formatted(
              nextDelivery.location(), nextEta.getSeconds() / 60);
      _emailGateway.send(nextDelivery.contactEmail(), "Your delivery will arrive soon", message);
    }
  }
}

File: kotlin/src/main/kotlin/kata/DeliveryController.kt
package kata

import java.time.Duration

class DeliveryController(val deliverySchedule: List<Delivery>) {

    private val emailGateway = EmailGateway()
    private val mapService = MapService()

    fun updateDelivery(deliveryEvent: DeliveryEvent) {
        var nextDelivery: Delivery? = null
        for ((i, delivery) in deliverySchedule.withIndex()) {
            if (deliveryEvent.id == delivery.id) {
                delivery.arrived = true
                val timeDifference = Duration.between(delivery.timeOfDelivery, deliveryEvent.timeOfDelivery)
                if (timeDifference < Duration.ofMinutes(10)) {
                    delivery.onTime = true
                }
                delivery.timeOfDelivery = deliveryEvent.timeOfDelivery
                val message =
                    "Regarding your delivery today at ${delivery.timeOfDelivery}. How likely would you be to recommend this delivery service to a friend? Click <a href='url'>here</a>"
                emailGateway.send(delivery.contactEmail, "Your feedback is important to us", message)
                if (deliverySchedule.size > i + 1) {
                    nextDelivery = deliverySchedule[i + 1]
                }

                if (!delivery.onTime && deliverySchedule.size > 1 && i > 0) {
                    val previousDelivery = deliverySchedule[i - 1]
                    val elapsedTime = Duration.between(previousDelivery.timeOfDelivery, delivery.timeOfDelivery)
                    mapService.updateAverageSpeed(previousDelivery.location, delivery.location, elapsedTime)
                }
            }
        }

        if (nextDelivery != null) {
            val nextEta = mapService.calculateETA(deliveryEvent.location, nextDelivery.location)
            val message =
                "Your delivery to ${nextDelivery.location} is next, estimated time of arrival is in ${nextEta} minutes. Be ready!"
            emailGateway.send(nextDelivery.contactEmail, "Your delivery will arrive soon", message)

        }
    }
}
